### ISSUE TITLE
False positive: Main.hs (executable) does not recognise function defined in Lib.hs (library)

### ISSUE THREAD

**Issue by mgpwilson** [haskell/haskell-language-server#1822]
Hello, I am having trouble with erroneous error messages from the linter.

The steps I have taken are:
1. Created a new Haskell Stack project with `stack new <project-name>`
2. Used `stack setup`
3. Created a function in Lib.hs and then used that function in Main.hs

Below are screenshots of the error I see.

![image](https://user-images.githubusercontent.com/9961083/117804358-9791b000-b24f-11eb-9704-b77b10a83863.png)
![image](https://user-images.githubusercontent.com/9961083/117804387-9fe9eb00-b24f-11eb-9ce4-697e452046a4.png)

Whenever I use `stack run`, the program compiles and runs as expected.

**Comment by googleson78** [comment]
I think this is also related to https://github.com/haskell/haskell-language-server/issues/735  (same root issue?)

**Comment by jneira** [comment]
> 
> 
> I think this is also related to #735 (same root issues?)

Yeah, thanks for linking the previous issue, not totally sure though

**Comment by mgpwilson** [comment]
Hi there,

Adding a `hie.yaml` file with the contents
```
cradle:
  stack:
```
seemed to fix the issue for me.

**Comment by googleson78** [comment]
> > I think this is also related to #735 (same root issues?)
> 
> Yeah, thanks for linking the previous issue, not totally sure though

I was linking it because I constantly get this "staleness" issue - if I add a new export to something in `lib`, hls won't detect in in another component, until I manually rebuild `lib`. And it seems related to me, because they're both related to intercomponent dependencies of some sort.

**Comment by jneira** [comment]
> 
> 
> Hi there,
> 
> Adding a `hie.yaml` file with the contents
> 
> ```
> cradle:
>   stack:
> ```
> 
> seemed to fix the issue for me.

Nice to see you can workaround the issue, have you got the opportunity to check if removing `hie.yaml` the issue is reproduced again?

**Comment by mgpwilson** [comment]
Hi @jneira

When I remove the `hie.yaml` file I get the following error:

![image](https://user-images.githubusercontent.com/9961083/117974304-5370df00-b325-11eb-9ea3-2296399a67e0.png)

When I reload developer window, that error is gone and the former error is reproduced:

![image](https://user-images.githubusercontent.com/9961083/117974576-a2b70f80-b325-11eb-9a99-5119e51c2060.png)

![image](https://user-images.githubusercontent.com/9961083/117974609-a9de1d80-b325-11eb-9bf7-bebaebe45be1.png)

**Comment by jneira** [comment]
Thanks for confirming it, @fendor does that behaviour make sense for you?

**Comment by mgpwilson** [comment]
> Hi, thanks for the bug report. Unfortunately the editor not always reflect changes on the fly, does the error persist after closing the editor, execute `stack build` in a console and reopening the editor?

Sorry I didn't reply to this before. The error did not persist after following the steps you listed. Each time I reloaded the developer window, the linter would recognise changes that were present at reloading.

**Comment by jneira** [comment]
> The error did not persist after following the steps you listed. Each time I reloaded the developer window, the linter would recognise changes that were present at reloading.

Ok, it would be great to distinguish cases, to help trace the cause, so taking in account your last comment: with `hie.yaml` you dont need to restart and execute `stack build`?

**Comment by fendor** [comment]
@jneira No, not really. Maybe the generated `hie.yaml` file is interesting?

**Comment by jneira** [comment]
It is the simpler stack one: https://github.com/haskell/haskell-language-server/issues/1822#issuecomment-839652383

**Comment by fendor** [comment]
but it doesn't work without it, `implicit-hie` generates a more elaborated `hie.yaml` than that.

**Comment by mgpwilson** [comment]
Okay so I've made a new project to try to provide as much information as possible:

Initial steps:
1. Created new package with `stack new ninety-nine-problems`
2. Opened code in folder `ninety-nine-problems`

After modifying Lib.hs, the following error is in Main.hs

![image](https://user-images.githubusercontent.com/9961083/118139205-09f4c280-b3ff-11eb-99b0-d4bb1d44199b.png)

3. Ran `stack setup`. Above error persists

4. Run `stack build`. Above error is gone!

5. Define `printSomething` in Lib.hs and use it in the main function of Main.hs. The below error appears.

![image](https://user-images.githubusercontent.com/9961083/118140669-7cb26d80-b400-11eb-9af1-2bdb4a329893.png)

6. Run `stack run`. Code compiles as expected. Error persists.
7. Reload developer window. Error is gone!
8. Define a new function `printSomethingElse` and use it in main. Variable not in scope error reappears.
9. Reload developer window without running `stack build`. Error persists.
10. Create `hie.yaml` with same contents as in https://github.com/haskell/haskell-language-server/issues/1822#issuecomment-839652383. Error persists.
11. Reload developer window. Error is gone.
12. Define new function `printSomethingElseElse` in Lib.hs and use it in Main.hs. No error.

Apologies, I'm not too used to reporting bugs but I hope that provides some use. Let me know if there's anything else you specifically want me to test.

**Comment by fendor** [comment]
This is an awesome bug report! Very detailed and great to follow. I will try to reproduce it locally as well.

**Comment by fendor** [comment]
Can reproduce as described.

Interesting finds:
With the implicit-cradle, we see the following log:
```
Typechecking reverse dependencies for NormalizedFilePath "/home/hugin/Documents/haskell/ninety-nine/src/Lib.hs": Just []
```
Which is wrong, `Main` is one of the reverse dependencies.

Compare to the logs with `hie.yaml`:
```
Typechecking reverse dependencies for NormalizedFilePath "/home/hugin/Documents/haskell/ninety-nine/src/Lib.hs": Just [NormalizedFilePath "/home/hugin/Documents/haskell/ninety-nine/app/Main.hs"]
```

The ghc options produced by `hie-bios` are completely identical.

I blindly guess that this is a bug in ghcide session loading.

**Comment by jneira** [comment]
The issue was reproduced by @dharmatech in #2058, log here: https://pastebin.com/hGtJMiC7, cabal file here: https://github.com/haskell/haskell-language-server/issues/2058#issuecomment-890863653

**Comment by santiweight** [comment]
Heya, my intention is not to clog - I know all of you are busy and put great work into HLS.

I'll just give honest feedback that this issue makes HLS very very abrasive for most usages. Almost any project has some form of test->lib dependency, and having to reload the entire server when coding almost any small change is a pretty big inconvenience.

I think you all know this, but I just want to check in with my experience, given that this issue seems to have persisted for some 8ish months now (there is a similar but older issue that parallelsl this one). If you have anywhere to point me to, I would be happy to take a look and see if I can do some debugging. Alternatively, if this issue is for more seasoned HLS contributors (which I suspect is the case), then I would be happy to start a git "bounty" or buy someone 100 coffees to help speed things along :), or help someone however that help is wanted...

**Comment by jneira** [comment]
@santiweight absolutely agree with you, this bug makes hls almost unusable with stack
we found that adding a simple stack explicit cradle seemed to workaround the issue, does it for you too?
I would ask @Ailrun how he deals with this issue, as he is using stack iirc

**Comment by Ailrun** [comment]
I have a shortcut key for restart in my setup, and just press the key when I cross the section border.

**Comment by santiweight** [comment]
Thanks for the response. I have an explicit stack cradle outlined like this:

```
cradle:
   stack:
     - path: "./src"
       component: "poker-base:lib"
     - path: "./test"
       component: "poker-base:test:all-tests"```

@Ailrun that's also my settle solution for the time being.

**Comment by fendor** [comment]
I spent a couple of minutes on it again, and it honestly seems like a stack bug.

E.g. when I change something in `Lib.hs`, then:
``` sh 
> stack repl ninety-nine-problems:exe:ninety-nine-problems-exe
Using main module: 1. Package `ninety-nine-problems' component ninety-nine-problems:exe:ninety-nine-problems-exe with main-is file: /home/baldr/Documents/haskell/ninety-nine-problems/app/Main.hs
ninety-nine-problems> configure (lib + exe)
Configuring ninety-nine-problems-0.1.0.0...
ninety-nine-problems> initial-build-steps (lib + exe)
The following GHC options are incompatible with GHCi and have not been passed to it: -threaded
Configuring GHCi with the following packages: ninety-nine-problems
GHCi, version 8.10.4: https://www.haskell.org/ghc/  :? for help
Loaded GHCi configuration from /home/baldr/.ghci
[1 of 2] Compiling Main             ( /home/baldr/Documents/haskell/ninety-nine-problems/app/Main.hs, interpreted )

/home/baldr/Documents/haskell/ninety-nine-problems/app/Main.hs:6:8: error:
    • Variable not in scope: someFunc2 :: IO ()
    • Perhaps you meant ‘someFunc’ (imported from Lib)
  |
6 | main = someFunc2
  |        ^^^^^^^^^
Failed, no modules loaded.

<no location info>: error:
    Could not load module ‘Paths_ninety_nine_problems’
    it is a hidden module in the package ‘ninety-nine-problems-0.1.0.0’
Loaded GHCi configuration from /run/user/1000/haskell-stack-ghci/80188eea/ghci-script
```

Which should not have happened, it should have rebuilt the library component.

Admittedly, it should work when HLS opens the Library component, but one issue after another.

**Comment by fendor** [comment]
And from 

``` sh
> hie-bios debug src/Lib.hs
...
GHC options:           -i -odir=/home/baldr/Documents/haskell/ninety-nine-problems/.stack-work/odir -hidir=/home/baldr/Documents/haskell/ninety-nine-problems/.stack-work/odir -hide-all-packages -i/home/baldr/Documents/haskell/ninety-nine-problems/.stack-work/dist/x86_64-linux-nix/Cabal-3.2.1.0/build -i/home/baldr/Documents/haskell/ninety-nine-problems/src -i/home/baldr/Documents/haskell/ninety-nine-problems/.stack-work/dist/x86_64-linux-nix/Cabal-3.2.1.0/build/autogen -i/home/baldr/Documents/haskell/ninety-nine-problems/.stack-work/dist/x86_64-linux-nix/Cabal-3.2.1.0/build/global-autogen -stubdir=/home/baldr/Documents/haskell/ninety-nine-problems/.stack-work/dist/x86_64-linux-nix/Cabal-3.2.1.0/build -I/nix/store/vkyg60kc8bzn089vjmadwwx0iaa4l440-ghc-8.10.4/include -I/nix/store/vygn2385bngrgfdagb3gds6bxfp5h4dc-git-2.31.1/include -I/nix/store/74kv08wjf06ifgk4dxrnra4qhzr5s1w4-gcc-wrapper-10.3.0/include -I/nix/store/wm9fdwbqvawsr5i66hp6kb9vfznapdn7-gmp-6.2.1-dev/include -L/nix/store/vkyg60kc8bzn089vjmadwwx0iaa4l440-ghc-8.10.4/lib -L/nix/store/vygn2385bngrgfdagb3gds6bxfp5h4dc-git-2.31.1/lib -L/nix/store/74kv08wjf06ifgk4dxrnra4qhzr5s1w4-gcc-wrapper-10.3.0/lib -L/nix/store/qkc31n1f1bs1l8m7k4fxyp4ykidy2cjr-gmp-6.2.1/lib -package-id=base-4.14.1.0 -optP-include -optP/home/baldr/Documents/haskell/ninety-nine-problems/.stack-work/ghci/ebbf0c21/cabal_macros.h -ghci-script=/run/user/1000/haskell-stack-ghci/a6a94610/ghci-script -package-db /home/baldr/Documents/haskell/ninety-nine-problems/.stack-work/install/x86_64-linux-nix/0cddea7c3f96a91112b31619351e85e026cd5fa5d73c071232a9ca415db9ac3c/8.10.4/pkgdb -package-db /home/baldr/.stack/snapshots/x86_64-linux-nix/0cddea7c3f96a91112b31619351e85e026cd5fa5d73c071232a9ca415db9ac3c/8.10.4/pkgdb -package-db /nix/store/vkyg60kc8bzn089vjmadwwx0iaa4l440-ghc-8.10.4/lib/ghc-8.10.4/package.conf.d
...
```

A bit much, but the important bit is that we have no flag such as `-this-unit-id`. Thus, HLS can not know what the component is named, and consequentially HLS can not handle the library component dependency of the executable. Therefore, reloading fails, as HLS is not aware of the local dependency.

Unfortunately, the final verdict is, thus, `cant fix here`, we need to patch stack.

**Comment by jneira** [comment]
@fendor thanks a lot to take a look, your wisdom was needed here :slightly_smiling_face: 

A ticket in the stack issue tracker would be really nice (a new or existing one) to continue make progress on this.
Otoh, this has been a regression?, does it oowrk in hls at some point?

**Comment by Ailrun** [comment]
AFAIK it haven't, at least for HLS (not sure about HIE, but I doubt it was the case)

**Comment by jneira** [comment]
> Unfortunately, the final verdict is, thus, cant fix here, we need to patch stack.

@fendor Thanks again for the analysis. For completeness and to help move upstream the fix, we have two issues:

* `stack repl exec-comp` does not rebuild the lib component if it has changed. This one is easy to reproduce and has a clear manifestation in a direct use of stack itself
* stack does not pass `--this-unit-id` when doing a `stack repl` (or never?) and hls cant use it to identify the library component (theoretically there could be several ibs in a package). Not sure if that is observable in a direct use of stack.

do we need to fix both things upstream to fix this issue or only the second one?

**Comment by fendor** [comment]
I *think* I remember seeing an issue for the former already... but don't have it at hand.

Well, both are required for a smooth experience, otherwise you would run into an error when you only open the executable which gets fixed once you open the library component.
Ideally, we fix both, and  `s-b-i` will be helpful here as well

**Comment by jneira** [comment]
Yeah we have already an issue about the former in hls: #318 linked with the upstream one: https://github.com/commercialhaskell/stack/issues/4616
So we have to open or find another one about the `--this-unit-id` thing. Not sure how to fill it tbh.

**Comment by jneira** [comment]
I think this is essentially https://github.com/haskell/haskell-language-server/issues/366, there we have the workarounds and linked upstream stack issues

**Comment by fendor** [comment]
I re-open this, since I don't feel like these issues are truly the same, they seem to have different root causes and present themselves differently.
